src/coreComponents/finiteVolume/SurfaceElementStencil.hpp

```diff
@@ -327,59 +389,104 @@ SurfaceElementStencilWrapper::
                   real64 ( & weight )[maxNumConnections][2],
                   real64 ( & dWeight_dVar )[maxNumConnections][2] ) const
 {
-
-  real64 sumOfTrans = 0.0;
-  for( localIndex k=0; k<numPointsInFlux( iconn ); ++k )
-  {
-    localIndex const er  =  m_elementRegionIndices[iconn][k];
-    localIndex const esr =  m_elementSubRegionIndices[iconn][k];
-    localIndex const ei  =  m_elementIndices[iconn][k];
-
-    sumOfTrans += coefficient[er][esr][ei][0][0] * m_weights[iconn][k];
-  }
-
   localIndex k[2];
   localIndex connectionIndex = 0;
   for( k[0]=0; k[0]<numPointsInFlux( iconn ); ++k[0] )
   {
     for( k[1]=k[0]+1; k[1]<numPointsInFlux( iconn ); ++k[1] )
     {
-      localIndex const er0  =  m_elementRegionIndices[iconn][k[0]];
-      localIndex const esr0 =  m_elementSubRegionIndices[iconn][k[0]];
-      localIndex const ei0  =  m_elementIndices[iconn][k[0]];
-
-      localIndex const er1  =  m_elementRegionIndices[iconn][k[1]];
-      localIndex const esr1 =  m_elementSubRegionIndices[iconn][k[1]];
-      localIndex const ei1  =  m_elementIndices[iconn][k[1]];
+      real64 halfWeight[2];
+      real64 dHalfWeight_dVar[2];
+      const localIndex elem[2] = {k[0], k[1]};
+      for( localIndex ielem=0; ielem < 2; ++ielem )
+      {
 
-      real64 const t0 = m_weights[iconn][0] * coefficient[er0][esr0][ei0][0][0]; // this is a bit insane to access perm
-      real64 const t1 = m_weights[iconn][1] * coefficient[er1][esr1][ei1][0][0];
+        localIndex const er  =  m_elementRegionIndices[iconn][elem[ielem]];
+        localIndex const esr =  m_elementSubRegionIndices[iconn][elem[ielem]];
 
-      real64 const harmonicWeight   = t0*t1 / sumOfTrans;
-      real64 const arithmeticWeight = 0.25 * (t0+t1);
+        computeWeightsBase( iconn, elem, ielem,
+                            coefficient[er][esr], dCoeff_dVar[er][esr],
+                            halfWeight[ielem], dHalfWeight_dVar[ielem] );
 
-      real64 const value = m_meanPermCoefficient * harmonicWeight + (1 - m_meanPermCoefficient) * arithmeticWeight;
+      }
 
-      weight[connectionIndex][0] = value;
-      weight[connectionIndex][1] = -value;
+      averageWeights( iconn, connectionIndex, halfWeight, dHalfWeight_dVar, weight, dWeight_dVar );
+      connectionIndex++;
 
-      real64 const dt0 = m_weights[iconn][0] * dCoeff_dVar[er0][esr0][ei0][0][0];
-      real64 const dt1 = m_weights[iconn][1] * dCoeff_dVar[er1][esr1][ei1][0][0];
+    }
+  }
+}
 
-      real64 dHarmonic[2];
-      dHarmonic[0] = ( dt0 * t1 * sumOfTrans - dt0 * t0 * t1 ) / ( sumOfTrans * sumOfTrans );
-      dHarmonic[1] = ( t0 * dt1 * sumOfTrans - dt1 * t0 * t1 ) / ( sumOfTrans * sumOfTrans );
+    GEOS_HOST_DEVICE
+    inline void
+    SurfaceElementStencilWrapper::
+    computeWeights( localIndex iconn,
```

Comment:
isn't this method the same for all stencils? I feel that the only method that needs to be specialized is `computeWeightsBase`.

Reply:
Not sure I understand here. Is that all stencil get a 
```
GEOS_HOST_DEVICE
inline void
EmbeddedSurfaceToCellStencilWrapper::
  computeWeights( localIndex iconn,
                  CoefficientAccessor< arrayView3d< real64 const > > const & coefficient,
                  CoefficientAccessor< arrayView3d< real64 const > > const & dCoeff_dVar,
                  real64 ( & weight )[1][2],
                  real64 ( & dWeight_dVar )[1][2] ) const;
``` 
and should then get the phase-indexed extension ? If so, it is unfortunately not `SurfaceElementStencil` 's case as it has `maxNumConnections` set to 6.

Otherwise there is quite some overloads of weight calcultation that take up to two variable deps.